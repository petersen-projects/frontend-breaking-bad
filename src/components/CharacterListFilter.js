import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';

import { setNameFilter, setCharacterStatusFilter } from '../actions/filters';
import { startGetAllCharactersFromWs } from '../actions/characters';

export const CharacterListFilter = () => {
   const name = useSelector(state => state.filters.name)
   const status = useSelector(state => state.filters.status)
   const dispatch = useDispatch()

   const onNameChange = (e) => {
      dispatch(setNameFilter(e.target.value))  
   }
   const onStatusFilterChange = (e) => {
      dispatch(setCharacterStatusFilter(e.target.value))
   }

   return (
      <div>
         <input
            type="text"
            placeholder="Search by character name..."
            value={name}
            onChange={onNameChange}
         />
         <select
            className=""
            value={status}
            onChange={onStatusFilterChange}
         >
            <option value="">Choose one Status</option>
            <option value="alive">Alive</option>
            <option value="deceased">Deceased</option>
         </select>
         <button>Search</button>
      </div>
   )
}

export default CharacterListFilter;