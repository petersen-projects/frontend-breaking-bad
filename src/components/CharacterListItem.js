import React from 'react';

export const CharacterListItem = ({
   name, age, img, status
}) => (
   <div className="list-row">
      {/* <div className="list-row__image-container">
      <img src={img}></img>
      </div> */}
      <div className="list-row__info">
         <div className="list-row__title">{name}</div>
         <div>Age: {age}</div>
         <div>Status: {status}</div>
      </div>
   </div>
);

export default CharacterListItem;