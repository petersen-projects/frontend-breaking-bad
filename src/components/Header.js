import React from 'react';

const Header = () => (
   <div className="header">
      <div className="header__title">
         Breaking Bad Characters
      </div>
      <div className="header__made-by">
         made by: Oggo Petersen
      </div>
   </div>
);

export default Header;