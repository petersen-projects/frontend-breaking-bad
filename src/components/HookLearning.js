import React, { useState, useEffect } from 'react';

export const HookLearning = () => {
   const [notes, setNotes] = useState([]);
   const [title, setTitle] = useState('');
   const [body, setBody] = useState('');

   useEffect(() => {
      const notesData = JSON.parse(localStorage.getItem('notes3'));
      setNotes(notesData || []);
   }, []);

   useEffect(() => {
      localStorage.setItem("notes3", JSON.stringify(notes));
   }, [JSON.stringify(notes)])

   const setTitleText = (e) => {
      setTitle(e.target.value);
   }

   const setBodyText = (e) => {
      setBody(e.target.value);
   }


   const addNote = (e) => {
      e.preventDefault();
      setNotes([
         ...notes,
         {
            title,
            body
         }
      ]);
      setTitle("");
      setBody("");
   }

   const removeNote = (title) => {
      setNotes(notes.filter(note => note.title !== title));
   }

   return (
      <div>
         <h1>Notes</h1>
         {notes.map(note => (
            <Note key={note.title} note={note} removeNote={removeNote} />
         ))}
         <p>Add note</p>
         <form onSubmit={addNote}>
            <input type="text" value={title} onChange={setTitleText}/>
            <input type="text" value={body} onChange={setBodyText}/>
            <button>add note</button>
         </form>
      </div>
   );
};

const Note = ({ note, removeNote }) => {
   useEffect(() => {
      console.log('useEffect Note');

      return () => {
         console.log('Cleaning up effect!')
      }
   }, [])

   return (
      <div>
         <h3>{note.title}</h3>
         <p>{note.body}</p>
         <button onClick={() => removeNote(note.title)}>x</button>
      </div>
   )
}

// export const HookLearning = (props) => {
//    const [count, setCount] = useState(props.count)
//    const [text, setText] = useState('');

//    useEffect(() => {
//       console.log('run once');
//    }, []);

//    useEffect(() => {
//       console.log('useeffect ran');
//       document.title = count;
//    }, [text]);

//    const increment = () => {
//       setCount(count + 1);
//    };

//    const decrement = () => {
//       setCount(count - 1);
//    };

//    const reset = () => {
//       setCount(0);
//    };

//    const changeText = (e) => {
//       setText(e.target.value);
//    }

//    return (
//       <div>
//             <input type="text" value={text} onChange={changeText} /><br/>
//             <p>the current {text || 'count'}  is {count}</p>
//             <button onClick={increment}>+1</button>
//             <button onClick={decrement}>-1</button>
//             <button onClick={reset}>reset</button>
            
//       </div>
//    )
// };

HookLearning.defaultProps = {
   count: 0
};

export default HookLearning;