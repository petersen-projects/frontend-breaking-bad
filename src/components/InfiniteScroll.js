import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';


export const InfiniteScroll = () => {

}

export default InfiniteScroll;
