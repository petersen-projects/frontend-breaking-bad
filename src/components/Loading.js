import React from 'react';

const Loading = () => (
   <div className="loader-container">
      <div className="loader">
         <div className="loader__icon"></div>
      </div>   
   </div>
);

export default Loading;