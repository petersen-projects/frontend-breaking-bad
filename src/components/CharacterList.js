import React, {
   useRef,
   useCallback,
   useEffect
} from 'react';
import { useSelector, useDispatch } from 'react-redux';

import {
   setCharactersOffset,
   resetCharactersOffset,
   startGetAllCharactersFromWs,
   setCharactersFilteredData,
   resetCharactersFilteredData,
   setCharactersNumberOfPages
} from '../actions/characters';
import { selectCharacters, getNumberOfPages } from '../selectors/characters';
import CharacterListItem from './CharacterListItem';
import Loading from './Loading';

const CharacterList = () => {
   const dispatch = useDispatch()

   const data = useSelector(state => state.characters.data)
   const filteredData = useSelector(state => state.characters.filteredData)
   const fetching = useSelector(state => state.characters.fetching)
   const numberOfPages = useSelector(state => state.characters.numberOfPages)
   const offset = useSelector(state => state.characters.offset)
   const limit = useSelector(state => state.characters.limit)
   const name = useSelector(state => state.filters.name)
   const status = useSelector(state => state.filters.status)

   useEffect(() => {
      dispatch(startGetAllCharactersFromWs())
   }, [])

   useEffect(() => {
      updateFilteredData(data)
   }, [data])

   useEffect(() => {
      if (offset > 0) {
         resetOffset()
      } else {
         refreshFilteredData()   
      }
   }, [name, status])

   useEffect(() => {
      if (offset > 0) {
         paginateData()
      } else {
         refreshFilteredData()
      }
   }, [offset])

   const observer = useRef()
   const lastCharacterRef = useCallback(node => {
      if (Math.ceil(filteredData.length / limit) >= numberOfPages) return
      if (observer.current) observer.current.disconnect()

      observer.current = new IntersectionObserver(entries => {
         if (entries[0].isIntersecting) {
            dispatch(setCharactersOffset(offset + 1))
         }
      })

      if (node) observer.current.observe(node)
   })

   const updateFilteredData = (() => {
      if (data.length) {
         dispatch(
            setCharactersFilteredData(
               selectCharacters({data, limit})
            )
         )
         dispatch(
            setCharactersNumberOfPages(
               getNumberOfPages({data, limit})
            )
         )
      }
   })

   const resetOffset = (() => {
      dispatch(resetCharactersOffset())
   })

   const refreshFilteredData = (() => {
      dispatch(
         resetCharactersFilteredData()
      )
      dispatch(
         setCharactersFilteredData(
            selectCharacters({data, limit, offset, name, status})
         )
      )
      dispatch(
         setCharactersNumberOfPages(
            getNumberOfPages({data, limit, offset, name, status})
         )
      )
   })

   const paginateData = () => {
      dispatch(
         setCharactersFilteredData(
            selectCharacters({
               data, offset, name, status
            })
         )
      )
   }

   return (
      <div className="list-body">
         { fetching == true ? <Loading /> : null}
         {
            filteredData.length === 0 ? (
               <div className="list-body__no-data-message">No data available</div>
            ) : (
               filteredData.map((character, index) => {
                  if (filteredData.length === index + 1) {
                     return (
                        <div ref={lastCharacterRef} key={index}>
                           <CharacterListItem {...character} />
                        </div>
                     )
                  } else {
                     return <CharacterListItem key={index} {...character} />
                  }
                     
               })
            )
         }
      </div>
   )
}

export default CharacterList;
