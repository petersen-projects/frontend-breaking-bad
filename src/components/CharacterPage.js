import React from 'react';

import CharacterList from './CharacterList';
import CharacterListFilter from './CharacterListFilter';
import Header from './Header';

export const CharacterPage = () => (
   <div className="character-page-container">
      <Header />
      <CharacterListFilter />
      <CharacterList />
   </div>
);

export default CharacterPage;