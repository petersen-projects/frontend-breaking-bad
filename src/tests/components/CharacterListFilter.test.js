import React from 'react';
import configureMockStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import { Provider } from 'react-redux';
import { mount } from 'enzyme';

import { CharacterListFilter } from '../../components/CharacterListFilter';

const mockStore = configureMockStore([thunk]);
const initialState = {
   characters: {
      data: [],
      filteredData: []
   },
   filters: {}
};

let setCharacterNameFilter, wrapper;

beforeEach(() => {
   setCharacterNameFilter = jest.fn();
   wrapper = mount(
      <Provider store={mockStore(initialState)}>
         <CharacterListFilter
            filters={{}}
            setCharacterNameFilter={setCharacterNameFilter}
         />
      </Provider>
   );
});

test('should render CharacterListFilter correctly', () => {
   expect(wrapper).toMatchSnapshot();
});
