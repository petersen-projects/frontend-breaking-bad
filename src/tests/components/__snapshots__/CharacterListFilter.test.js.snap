// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render CharacterListFilter correctly 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <CharacterListFilter
    filters={Object {}}
    setCharacterNameFilter={[Function]}
  >
    <div>
      <input
        onChange={[Function]}
        placeholder="Search by character name..."
        type="text"
      />
      <select
        className=""
        onChange={[Function]}
      >
        <option
          value=""
        >
          Choose one Status
        </option>
        <option
          value="alive"
        >
          Alive
        </option>
        <option
          value="deceased"
        >
          Deceased
        </option>
      </select>
      <button>
        Search
      </button>
    </div>
  </CharacterListFilter>
</Provider>
`;
